docs/developers/decisions/backend/0016-garbage-collector-design.md

```diff
@@ -0,0 +1,316 @@
+# Context and Problem Statement
+
+!!! note "This is a design document"
+
+    This document focuses more on how GC will be implemented in the application,
+    rather than the tradeoffs/benefits of having a GC in the first place, etc.
+
+As users interact with the application, creating and deleting loadouts, mods and more;
+there will be an ever-growing amount of files which are no longer needed on disk.
+
+Sources of unused files may include:
+
+- Removed loadouts.
+    - ***Example:*** After removing a loadout, some `LoadoutItems` backed by a file are no longer referenced in any loadout.
+- Trimmed/reset loadout history.
+    - ***Example:*** Removing loadout history will remove references to `LoadoutItems`, which will leave unreferenced files.
+
+## Design Overview
+
+The Garbage Collection mechanism is implemented using a `mark-and-sweep` approach.
+
+This can be broken down into the following steps:
+
+- [Parse all `Nx` archives used by the App](#parsing-nx-archives).
+    - The `Nx` archives contain the hashes of all files in its headers.
+    - This finds out all the files we've got packed.
+    - This gives us ***AllFileList (A)***
+- [Perform a `mark` phase through the DataStore.](#performing-the-mark-phase)
+    - Find all files that are in use.
+    - This is any `LoadoutItem` which is a file (`LoadoutItemWithTargetPath`) that is not marked as deleted (retracted).
+    - This gives us ***UsedFileList (B)***.
+- [Remove all elements in ***AllFileList (A)*** that are *not* in ***UsedFileList (B)***](#removing-unused-items).
+    - The removal process involves `'repacking'` all archives which have had their files
+removed to exclude the files that are no longer needed.
+
+### Parsing Nx Archives
+
+!!! note "This process involves identifying all archives the App has available, and which files are contained within them."
```

Comment:
Don't need to do this, we can instead do `db.Datoms(ArchivedFile.Hash)` and that will return a datom for each archived file. Bonus: the results will be sorted by hash

Reply:
[Note: I went a bit overboard with the response here, but it may be good to include in ADR, because this is a good question; and I was feeling a little bit passionate]

--------

I technically don't need to do this step to find out which files are not in use.

*However*, it is more efficient for me to do so.
With `db.Datoms(ArchivedFile.Hash)`, I would need to do the following:

- Make 1 DB query to get the container name for every hash to determine the `Nx`.
  - And for every hash, check all locations in `DataModelSettings.ArchiveLocations` to see if the file exists.
      - Which basically means opening a file handle and seeing if it works.
      - In most cases there's only 1 `ArchiveLocation`, so usually 1 time per file.
  - i.e. Same as `NxFileStore.TryGetLocation`.

If I instead read the archive headers to get this information; the following applies:

- Scan a folder for all archives.
    - On Windows, around 450 file (archive) entries per millisecond. Uses `NtQueryDirectoryFileEx` under the hood.
    - Can scan multiple folders in parallel, and that does scale.
- Open each archive, parse the header. Usually just first 4K page for >90% of the mods.
- Grab the hashes of multiple files from the Nx header.

In this case, a handle is also opened for every archive file, to read the header.
The part that is slow on Windows is unfortunately opening that handle, and that is something we do in both situations.

In the datoms case, I technically would need to open a file multiple times; but I can avoid that with a `fileExistsCache` like I did in `NxFileStore`.

------

So what it really comes down to is asking the following:

1. Whether querying container name `entry.Container` for every datom `db.Datoms(ArchivedFile.Hash)` is faster than `NtQueryDirectoryFileEx`.

2. Whether reading Hash Datoms from DB is faster than reading Nx archive headers.

For both steps, I think they're about equal.

For step `1`, reading the MFT on an NTFS volume is mostly a sequential operation, with some occasional fragmentation leading to jumping around. Reading the DB on the other hand involves reading from at least 2 places for the raw data, and an index lookup (3). Both approaches also open file handles, so we can cancel that part out.

For `2`, they're also about equal; assuming the datom data is laid out sequentially in RocksDB; then both the `FileEntry` structure (20 bytes) and MnemonicDB's Hash Entry [16 byte KeyPrefix + 8 byte Value] are about equal in size; and sequential.

At which point, you may ask; `why not just use the DB?`. The answer is very simple.

Regardless of the approach, I need to parse the Nx header regardless in order to perform the repacking step. That is unavoidable. This is because I need metadata such as `Chunk Size`, `Block Sizes` (and thus offsets), `Block Compressions` etc. Because the archive headers also contain the hashes, simply pulling the hashes from the pre-parsed headers directly from RAM is faster than firing DB calls.

Reply:
`Whether querying container name entry.Container for every datom db.Datoms(ArchivedFile.Hash) is faster than NtQueryDirectoryFileEx`

This is where set logic comes into play, you can have multiple sets of data in memory and link them up as needed. For example you could scan for all the file names (ignoring their size and contents) first that tells you how to filter the `ArchivedFileContainer` objects. From there you have a sorted set that you can use to probe into when figuring out what files reference what archives. 

In general though I think we should assume that all ArchivedFileContainers will exit if the DB says they do, going in and randomly deleting .nx archives is not supported behavior of the app. 

I think what I'm more concerned about here is that this ticket is dragging on, and we have some pretty complex logic here when the simpler approach could likely be nearing completion. I'd like to find ways of making this code at least testable soon and then go back and add microoptimizations as we figure out what's needed. Making any arguments about memory usage on something that isn't even testable yet is a bit out-of-order on the design. 

We can hook this into the UI to run as part of unmanaging a game, and then we have a "good enough" solution for now: whenever you delete the last loadout of a game, we run GC to delete any backed up files you no longer need.